vms/platformvm/utxo/camino_locked.go

```diff
@@ -707,6 +935,301 @@ func (h *handler) VerifyLockUTXOs(
+			// is not locked - this is invalid.
+			return nil, errLockedFundsNotMarkedAsLocked
+		}
+
+		needKeys := true
+		consumedAmount := in.Amount()
+
+		// calculating and checking unlockableAmount
+		// checking if deposit is expired and don't need keys
+
+		if isDeposited {
+			// we didn't calculate depositUnlocks (unlockableAmount, need keys)
+			// for this deposit before, so we'r calculating it here for the first time
+			depUnlock, ok := depositUnlocks[lockIDs.DepositTxID]
+			if !ok {
+				deposit, err := chainState.GetDeposit(lockIDs.DepositTxID)
+				if err != nil {
+					return nil, err
+				}
+
+				depositOffer, err := chainState.GetDepositOffer(deposit.DepositOfferID)
+				if err != nil {
+					return nil, err
+				}
+
+				depUnlock = &depositUnlock{
+					// if deposit isn't expired, than it could only be spended by its owner
+					needKeys: !deposit.IsExpired(depositOffer, currentTimestamp),
+					unlockableAmount: deposits.UnlockableAmount(
+						deposit,
+						depositOffer,
+						currentTimestamp,
+					),
+				}
+
+				depositUnlocks[lockIDs.DepositTxID] = depUnlock
+			}
+
+			needKeys = depUnlock.needKeys
+
+			newAmount, err := math.Add64(depUnlock.consumedAmount, consumedAmount)
+			if err != nil {
+				return nil, err
+			}
+			depUnlock.consumedAmount = newAmount
+		}
+
+		// verifying transfer
+
+		if needKeys {
+			if index >= len(creds) {
+				return nil, errInputsCredentialsMismatch
+			}
+
+			// Verify that this tx's credentials allow [in] to be spent
+			if err := h.fx.VerifyTransfer(tx, in, creds[index], out); err != nil {
+				return nil, fmt.Errorf("failed to verify transfer: %w", err)
+			}
+		}
+
+		// filling consumed map
+
+		ownerID, err := GetOwnerID(out)
+		if err != nil {
+			return nil, err
+		}
+
+		consumedOwnerAmounts, ok := consumed[ownerID]
+		if !ok {
+			consumedOwnerAmounts = make(map[ids.ID]uint64)
+			consumed[ownerID] = consumedOwnerAmounts
+		}
+
+		newAmount, err := math.Add64(consumedOwnerAmounts[lockIDs.BondTxID], consumedAmount)
+		if err != nil {
+			return nil, err
+		}
+		consumedOwnerAmounts[lockIDs.BondTxID] = newAmount
+	}
+
+	// iterating over outs, checking produced amounts with consumed map
+	// filling deposit produced amounts
+
+	for _, output := range outs {
+		out := output.Out
+		lockIDs := &locked.IDsEmpty
+		if lockedOut, ok := out.(*locked.Out); ok {
+			lockIDs = &lockedOut.IDs
+			out = lockedOut.TransferableOut
+		}
+
+		ownerID, err := GetOwnerID(out)
+		if err != nil {
+			return nil, err
+		}
+
+		producedAmount := out.Amount()
+		consumedAmount := uint64(0)
+		consumedOwnerAmounts, ok := consumed[ownerID]
+		if ok {
+			consumedAmount = consumedOwnerAmounts[lockIDs.BondTxID]
+		}
+
+		if consumedAmount < producedAmount {
+			return nil, fmt.Errorf(
+				"address %s produces %d and consumes %d for lockIDs %+v with lock '%s': %w",
+				ownerID,
+				producedAmount,
+				consumedAmount,
+				lockIDs.BondTxID,
```

Comment:
shouldn't that be DepositTxID ? or is bonded amounts somehow involved here?

Reply:
Its bond tx id, thats right - we need expect that all deposited inputs would be unlocked, so every depositTxID -> emptyID. But bond tx ids will persist, so we need to make sure that amount of tokens for every bond will remain the same